### PR TITLE
Add some basic runtime type validation in Onyx

### PR DESCRIPTION
**Pull request by dominictb**
### Details

Add some basic runtime type validation in Onyx for merge, set and mergeCollection operation

### Related Issues

https://github.com/Expensify/App/issues/39852

### Automated Tests

Test 1: For set operation: https://github.com/Expensify/react-native-onyx/blob/9ee724140cf4d78ce8d298545fa8445067a4f192/tests/unit/onyxTest.js#L75

Test 2: For merge operation: https://github.com/Expensify/react-native-onyx/blob/9ee724140cf4d78ce8d298545fa8445067a4f192/tests/unit/onyxTest.js#L117

Test 3: For `mergeCollection`: https://github.com/Expensify/react-native-onyx/blob/9ee724140cf4d78ce8d298545fa8445067a4f192/tests/unit/onyxTest.js#L434
   - Do not update the key if the value is not compatible with the existing value (array vs non-array)
   - Do not notify the subscriber's callback on the incompatible update.

### Manual Tests

N/A

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/react-native-onyx/assets/165644294/e4737868-d9c4-4b8c-af18-21bdea418e02

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/react-native-onyx/assets/165644294/c04b147e-6a0f-496d-a18b-b08481b31588

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/react-native-onyx/assets/165644294/a66c1482-1055-462c-874d-04052165d646

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/react-native-onyx/assets/165644294/177737b9-d955-4080-9242-5736bb248995

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/react-native-onyx/assets/165644294/31101ff4-ebb1-4c66-8441-41f31b414e18

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/react-native-onyx/assets/165644294/efd979e6-4840-4796-860b-77b94afad16c

</details>

### CODE REVIEW

**Comment by dominictb** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by dominictb** [comment]
recheck

**Comment by AndrewGable** [comment]
Assigned reviewers of https://github.com/Expensify/App/issues/39852

File: lib/logMessages.ts
Comment (roryabraham) [inline lib/logMessages.ts:2]: I think this error message could be improved to be more precise - "Trying to set an object to an array" or "Trying to merge and array into an object"

**Review by roryabraham — COMMENTED** [review]
@dominictb please complete the author checklist and perform some basic manual testing of this PR with a local build of E/App using these changes.

**Comment by dominictb** [comment]
> @dominictb please complete the author checklist and perform some basic manual testing of this PR with a local build of E/App using these changes.

Will do it as soon as possible. Thanks. Feels free to review the rest of the PR

File: lib/Onyx.ts
Comment (roryabraham) [inline lib/Onyx.ts:239]: NAB but let's please add this blank line back

File: lib/Onyx.ts
Comment (roryabraham) [inline lib/Onyx.ts:319]: Since this is a boolean, let's call it `isCompatible` please

```suggestion
                const {isCompatible, existingValueType, newValueType} = utils.checkCompatibilityWithExistingValue(change, existingValue);
```

Same for all the other usages

**Review by roryabraham — CHANGES_REQUESTED** [review]
Otherwise LGTM 👍

**Review by roryabraham — DISMISSED** [review]

**Comment by mananjadhav** [comment]
Reviewed it once, going through once again.

File: lib/utils.ts
Comment (mananjadhav) [inline lib/utils.ts:127]: Should we stick to string, or add a type.

```
type ValueType = 'array' | 'object' | 'other';

const existingValueType: ValueType = Array.isArray(existingValue) ? 'array' : (typeof existingValue === 'object' ? 'object' : 'other');


```

File: lib/utils.ts
Comment (mananjadhav) [inline lib/utils.ts:136]: Does harm if we return the type anyway?

File: lib/utils.ts
Comment (dominictb) [reply lib/utils.ts:127]: I would rather stick to a string, and we can use enum/constants if we need to implement some logic around that. It's more flexible

File: lib/utils.ts
Comment (dominictb) [reply lib/utils.ts:136]: It doesn't, but since we do not use it in the case of `incompatible`, I don't think returning the type in case of `compatible` make sense. Moreover, the compatibility check function currently shouldn't include the logic of getting the exact type of the values.

File: lib/utils.ts
Comment (mananjadhav) [reply lib/utils.ts:127]: I would also let @roryabraham weigh in on this.

**Review by mananjadhav — APPROVED** [review]
LGTM, pending one confirmation.

@roryabraham Do we do a checklist here? or when we bump the App PR with updated Onyx version?

**Comment by dominictb** [comment]
> LGTM, pending one confirmation.
> 
> @roryabraham Do we do a checklist here? or when we bump the App PR with updated Onyx version?

@mananjadhav some videos are uploaded in the PR description (local build). Feel free to check.

**Comment by mananjadhav** [comment]
Thanks @dominictb I did check those.

**Comment by roryabraham** [comment]
> Do we do a checklist here?

We should do a checklist here with a build of react-native-onyx linked against E/App dev. See [instructions in the README on how to do this.](https://github.com/Expensify/react-native-onyx?tab=readme-ov-file#development)

**Comment by dominictb** [comment]
uh oh, @mananjadhav anything I need to do to complete the PR? Thanks!

**Comment by mananjadhav** [comment]
No. I'll finish this checklist today.

**Comment by mananjadhav** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Review by mananjadhav — APPROVED** [review]

**Comment by mananjadhav** [comment]
@roryabraham All yours.

**Comment by roryabraham** [comment]
looks like reassure is just not set up correctly in Onyx?

**Review by roryabraham — COMMENTED** [review]
This code looks good, but @dominictb I think I need you to merge main into this branch so we can see if Reassure is passing or failing.

**Comment by dominictb** [comment]
@roryabraham I merged main. Can u check?

**Comment by mananjadhav** [comment]
@roryabraham Can you action the PR so that the workflows are executed?

**Comment by dominictb** [comment]
@roryabraham @mananjadhav let's merge this so I can bump the version in Expensify/App

**Comment by dominictb** [comment]
@roryabraham @mananjadhav any update before we can merge?

**Comment by roryabraham** [comment]
sorry for the delay. As this is your first PR in this repo I have to manually approve workflows to run, so that's why it was delayed.

**Review by roryabraham — APPROVED** [review]

**Pull request merged**
